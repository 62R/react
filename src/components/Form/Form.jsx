import { useState } from 'react';
import style from './Form.module.scss';

export const Form = ({ addMsg }) => {
  const [name, setName] = useState('');
  const [msg, setMsg] = useState('');
  const handleSubmit = (evt) => {
    evt.preventDefault();
    addMsg(name, msg);
    setMsg('');
  };
  return (
    <form onSubmit={handleSubmit} className={style.form}>
      <label htmlFor="name">Имя:</label>
      <input
        className={style.input}
        value={name}
        onInput={(evt) => setName(evt.target.value)}
        id="name"
      />
      <label htmlFor="msg">Сообщение:</label>
      <textarea
        className={style.msgArea}
        value={msg}
        onInput={(evt) => setMsg(evt.target.value)}
        id="msg"
        rows="5"
      />
      <button className={style.btn} disabled={!msg || !name}>
        Отправить
      </button>
    </form>
  );
};
