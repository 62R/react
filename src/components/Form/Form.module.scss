.form {
    display: flex;
    flex-direction: column;
    width: 100%;
}

.input {
    border: 1px solid cornflowerblue;
}

.msgArea {
    border: 1px solid cornflowerblue;
    resize: none;
}

.btn {
    margin-top: 10px;
    padding-top: 5px;
    padding-bottom: 5px;
    background-color: transparent;
    border: 1px solid cornflowerblue;
}