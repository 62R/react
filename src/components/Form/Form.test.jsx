import { Form } from './Form';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import userEvent from '@testing-library/user-event';

const mockHandler = jest.fn();
let inputs = [];

beforeEach(() => {
  render(<Form addMsg={mockHandler} />);
  inputs = screen.getAllByRole('textbox');
});

describe('Form', () => {
  it('render component with snapshot', () => {
    const { asFragment } = render(<Form addMsg={mockHandler} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('button is disabled, if form empty', () => {
    expect(screen.getByRole('button')).toBeDisabled();
  });

  it('dont submit when name not enter', async () => {
    await userEvent.type(inputs[1], 'Hello, world!');
    await userEvent.click(screen.getByRole('button'));

    expect(mockHandler).toHaveBeenCalledTimes(0);
  });

  it('dont submit when message not enter', async () => {
    await userEvent.type(inputs[0], 'Name');
    await userEvent.click(screen.getByRole('button'));

    expect(mockHandler).toHaveBeenCalledTimes(0);
  });

  it('submit when form complited', async () => {
    await userEvent.type(inputs[0], 'Name');
    await userEvent.type(inputs[1], 'Hello, world!');
    await userEvent.click(screen.getByRole('button'));

    expect(mockHandler).toHaveBeenCalledTimes(1);
  });

  it('clear message after submit', async () => {
    await userEvent.type(inputs[0], 'Name');
    await userEvent.type(inputs[1], 'Hello, world!');
    await userEvent.click(screen.getByRole('button'));

    expect(mockHandler).toHaveBeenCalledTimes(1);
    expect(inputs[1].value).toBe('');
  });

  it('name is not empty after submit', async () => {
    await userEvent.type(inputs[0], 'Name');
    const name = inputs[0].value;
    await userEvent.type(inputs[1], 'Hello, world!');
    await userEvent.click(screen.getByRole('button'));

    expect(mockHandler).toHaveBeenCalledTimes(1);
    expect(inputs[0].value).toBe(name);
  });
});
