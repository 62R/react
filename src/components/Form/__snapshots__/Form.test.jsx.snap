// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form render component with snapshot 1`] = `
<DocumentFragment>
  <form
    class="form"
  >
    <label
      for="name"
    >
      Имя:
    </label>
    <input
      class="input"
      id="name"
      value=""
    />
    <label
      for="msg"
    >
      Сообщение:
    </label>
    <textarea
      class="msgArea"
      id="msg"
      rows="5"
    />
    <button
      class="btn"
      disabled=""
    >
      Отправить
    </button>
  </form>
</DocumentFragment>
`;
